pilot/pkg/model/telemetry.go

```diff
@@ -0,0 +1,146 @@
+}
+
+// GetTelemetries returns the Telemetry configurations for the given environment.
+func GetTelemetries(env *Environment) (*Telemetries, error) {
+	telemetries := &Telemetries{
+		NamespaceToTelemetries: map[string][]Telemetry{},
+		RootNamespace:          env.Mesh().GetRootNamespace(),
+	}
+
+	fromEnv, err := env.List(collections.IstioTelemetryV1Alpha1Telemetries.Resource().GroupVersionKind(), NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+	sortConfigByCreationTime(fromEnv)
+	for _, config := range fromEnv {
+		telemetry := Telemetry{
+			Name:      config.Name,
+			Namespace: config.Namespace,
+			Spec:      config.Spec.(*tpb.Telemetry),
+		}
+		telemetries.NamespaceToTelemetries[config.Namespace] =
+			append(telemetries.NamespaceToTelemetries[config.Namespace], telemetry)
+	}
+
+	return telemetries, nil
+}
+
+func (t *Telemetries) EffectiveTelemetry(namespace string, workload labels.Collection) *tpb.Telemetry {
+	if t == nil {
+		return nil
+	}
+
+	var effectiveSpec *tpb.Telemetry
+	if t.RootNamespace != "" {
+		effectiveSpec = t.namespaceWideTelemetry(t.RootNamespace)
+	}
+
+	if namespace != t.RootNamespace {
+		nsSpec := t.namespaceWideTelemetry(namespace)
+		if nsSpec != nil && effectiveSpec != nil {
+			effectiveSpec = shallowMerge(effectiveSpec, nsSpec)
+		}
+	}
+
+	for _, telemetry := range t.NamespaceToTelemetries[namespace] {
+		spec := telemetry.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			continue
+		}
+		selector := labels.Instance(spec.GetSelector().GetMatchLabels())
+		if workload.IsSupersetOf(selector) {
+			effectiveSpec = shallowMerge(effectiveSpec, spec)
+		}
+	}
+
+	return effectiveSpec
+}
+
+func (t *Telemetries) namespaceWideTelemetry(namespace string) *tpb.Telemetry {
+	for _, tel := range t.NamespaceToTelemetries[namespace] {
+		spec := tel.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			return spec
+		}
+	}
+	return nil
+}
+
+func shallowMerge(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	return shallowMergeTracing(parent, child)
+}
+
+func shallowMergeTracing(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	merged := parent.DeepCopy()
+
+	if len(child.GetTracing()) == 0 {
+		return merged
+	}
+
+	childCopy := child.DeepCopy()
+
+	if len(merged.GetTracing()) == 0 {
+		merged.Tracing = childCopy.Tracing
+		return merged
+	}
+
+	// only use the first Tracing for now (all that is suppported)
+	childTracing := childCopy.Tracing[0]
+	mergedTracing := merged.Tracing[0]
+	if len(childTracing.Providers) != 0 {
+		mergedTracing.Providers = childTracing.Providers
+	}
+
+	if len(childTracing.CustomTags) != 0 {
+		mergedTracing.CustomTags = childTracing.CustomTags
+	}
+
+	if childTracing.DisableSpanReporting != mergedTracing.DisableSpanReporting {
+		mergedTracing.DisableSpanReporting = childTracing.DisableSpanReporting
+	}
+
+	if childTracing.RandomSamplingPercentage != 0 {
+		mergedTracing.RandomSamplingPercentage = childTracing.RandomSamplingPercentage
```

Comment:
same here, should it be a wrapper type?

Reply:
Here 0 value is being treated specifically, but a wrapper would be cleaner. Yes

Reply:
We should make sure to change to a wrapper pre-1.10, as its easier to change before shipping